File: lab_binary_search_recursive.cpp
#include <iostream>
#include <array>
#include <vector>
using namespace std;

bool binary_search(vector<int> &sequence, int value, int start, int end)
{
	if (end >= start)
	{
		int mid = start + (end - start) / 2;

		// If the element is present at the middle itself
		if (sequence[mid] == value)  return mid;

		// If element is smaller than mid, then it can only be present
		// in left subarray
		if (sequence[mid] > value) return binary_search(sequence, value, start, mid - 1);

		// Else the element can only be present in right subarray
		return binary_search(sequence, value, mid + 1, end);
	}
	return false;
}


int main()
{
	int errflg = 0;
	const int size = 33;
	vector<int> sequence;
	sequence.resize(size);

	// search test values and results
	struct Test
	{
		int value;
		bool expected;

	};
	array<Test, 4> testing = { { { 96,true },{ 33, true },{ 31, false },{ 76, false } } };

	// populate sequence
	int count = 0;
	for (int &i : sequence)
	{
		i = count;
		count += 3;
	}

	// carry out the tests
	for (Test test : testing)
	{
		bool result = binary_search(sequence, test.value,0,32);

		cout << (result == test.expected ? "Passed" : "Failed") <<
			" search for " << test.value << " test, got " <<
			(result ? "true" : "false") << " expected " <<
			(test.expected ? "true" : "false") << endl;

		if (result != test.expected) errflg += 1;
	}

	return errflg;
}

File: lab_binary_search_recursive.py
#!/usr/bin/python3
import sys

def binary_search( sequence, value ):
    def recurse(first, last):
        mid = int((first + last) / 2 )
        if first > last:
            return False
        elif (sequence[mid] < value):
            return recurse(mid + 1, last)
        elif (sequence[mid] > value):
            return recurse(first, mid - 1)
        else:
            return True 

    return recurse(0, len(sequence)-1)


def binary_search2( sequence, value, low,high ):
            mid = int((low + high) / 2 )
            if low > high:
                return False
            elif (sequence[mid] < value):
                return binary_search2(sequence, value,mid + 1, high)
            elif (sequence[mid] > value):
                return binary_search2(sequence, value,low, mid - 1)
            else:
                return True 



def main():
    errflg = 0
    sequence = [ i for i in range(0,100,3) ]
    testing = [ (96,True), (33,True), (31,False), (76,False) ]

    for value, expected in testing:
        #result = binary_search( sequence, value )
        result = binary_search2( sequence, value ,0,len(sequence))

        print( "%s search for %d test, got %r expected %r" % \
            ("Passed" if result is expected else "Failed", value, result, expected) )

        if result is not expected:
            errflg += 1

    return errflg

if __name__ == '__main__':
    sys.exit( main() )
